ospatch/zypper_patch.go

```diff
@@ -67,39 +71,100 @@ func ZypperUpdateWithUpdate(withUpdate bool) ZypperPatchOption {
 	}
 }
 
-// ZypperPatchRunner returns a ZypperUpdateOption that specifies the runner.
-func ZypperPatchRunner(runner func(cmd *exec.Cmd) ([]byte, error)) ZypperPatchOption {
+// ZypperUpdateWithExcludes returns a ZypperUpdateOption that specifies
+// list of packages to be excluded from update
+func ZypperUpdateWithExcludes(excludes []string) ZypperPatchOption {
+	return func(args *zypperPatchOpts) {
+		args.excludes = excludes
+	}
+}
+
+// ZypperUpdateWithExclusivePatches returns a ZypperUpdateOption that specifies
+// list of exclusive packages to be updated
+func ZypperUpdateWithExclusivePatches(exclusivePatches []string) ZypperPatchOption {
+	return func(args *zypperPatchOpts) {
+		args.exclusivePatches = exclusivePatches
+	}
+}
+
+// ZypperUpdateDryrun returns a ZypperUpdateOption that specifies the runner.
+func ZypperUpdateDryrun(dryrun bool) ZypperPatchOption {
 	return func(args *zypperPatchOpts) {
-		args.runner = runner
+		args.dryrun = dryrun
 	}
 }
 
 // RunZypperPatch runs zypper patch.
 func RunZypperPatch(opts ...ZypperPatchOption) error {
 	zOpts := &zypperPatchOpts{
-		runner: defaultRunner,
+		excludes:         nil,
+		exclusivePatches: nil,
+		categories:       nil,
+		severities:       nil,
+		withOptional:     false,
+		withUpdate:       false,
 	}
 
 	for _, opt := range opts {
 		opt(zOpts)
 	}
 
-	args := zypperPatchArgs
-	if zOpts.withOptional {
-		args = append(args, "--with-optional")
+	zListOpts := []packages.ZypperListOption{
+		packages.ZypperListPatchCategories(zOpts.categories),
+		packages.ZypperListPatchSeverities(zOpts.severities),
+		packages.ZypperListPatchWithOptional(zOpts.withOptional),
+		// if there is no filter on category and severity,
+		// zypper fetches all available patch updates
 	}
-	if zOpts.withUpdate {
-		args = append(args, "--with-update")
+	patches, err := packages.ZypperPatches(zListOpts...)
+	if err != nil {
+		return err
 	}
-	for _, c := range zOpts.categories {
-		args = append(args, "--category="+c)
+
+	fPatches, fpkgs, err := runFilter(patches, zOpts.exclusivePatches, zOpts.excludes, zOpts.withUpdate)
+	return packages.ZypperInstall(fPatches, fpkgs)
```

Comment:
Add logging and dryrun return like we have with the other systems:
https://github.com/GoogleCloudPlatform/osconfig/blob/master/ospatch/apt_upgrade.go#L89

Reply:
done